lib/pos_consensus_controller.go

```diff
@@ -56,13 +56,70 @@ func (cc *ConsensusController) HandleFastHostStuffEmptyTimeoutBlockProposal(even
 	// 6. Broadcast the block to the network
 }
 
-func (cc *ConsensusController) HandleFastHostStuffVote(event *consensus.FastHotStuffEvent) {
-	// The consensus module has signaled that we can vote on a block. We construct and
-	// broadcast the vote here:
-	// 1. Verify that the block height we want to vote on is valid
-	// 2. Construct the vote message
-	// 3. Process the vote in the consensus module
-	// 4. Broadcast the timeout msg to the network
+// HandleFastHostStuffVote is triggered when FastHotStuffEventLoop has signaled that it wants to
+// vote on the current tip. This functions validates the vote signal, then it constructs the
+// vote message here.
+//
+// Steps:
+// 1. Verify that the event is properly formed.
+// 2. Construct the vote message
+// 3. Process the vote in the consensus module
+// 4. Broadcast the vote msg to the network
+func (cc *ConsensusController) HandleFastHostStuffVote(event *consensus.FastHotStuffEvent) error {
+	// Hold a read lock on the consensus controller. This is because we need to check the
+	// current view and block height of the consensus module.
+	cc.lock.Lock()
+	defer cc.lock.Unlock()
+
+	var err error
+
+	if !consensus.IsProperlyFormedVoteEvent(event) {
+		// If the event is not properly formed, we ignore it and log it. This should never happen.
+		return errors.Errorf("HandleFastHostStuffVote: Received improperly formed vote event: %v", event)
+	}
+
+	// Provided the vote message is properly formed, we construct and broadcast it in a best effort
+	// manner. We do this even if the consensus event loop has advanced the view or block height. We
+	// maintain the invariant here that if consensus connected a new tip and wanted to vote on it, the
+	// vote should be broadcasted regardless of other concurrent events that may have happened.
+	//
+	// The block acceptance rules in Blockchain.ProcessBlockPoS guarantee that we cannot vote more
```

Comment:
```suggestion
	// The block acceptance rules in Blockchain.processBlockPoS guarantee that we cannot vote more
```

Reply:
ProcessBlockPoS will be a public function in the long run :)